# Take-home Task: Job-Market Intelligence Assistant with RAG and Graph RAG

## Overview

Build a lightweight research assistant that answers questions about skills, roles, companies, and job-market trends from public job-posting data.

The task is designed to assess practical experience with:

- LLM application development
- Retrieval-Augmented Generation (RAG)
- Embeddings and vector search
- Knowledge-graph modelling
- Graph-aware retrieval / Graph RAG
- Evaluation, source attribution, and communication

**Expected effort:** approximately **6–8 hours**, including documentation and presentation preparation.
We value sound scope decisions and a clear, working prototype over a polished product or a complex infrastructure setup.

---

## Data

The data used is a subset of the following dataset:

- [LinkedIn Job Postings – Kaggle](https://www.kaggle.com/datasets/arshkon/linkedin-job-postings)

The dataset contains job-posting text and structured information that may include jobs, companies, skills, industries, locations, and related metadata.

You may enrich the dataset with a public skills/occupation taxonomy such as [O*NET](https://www.onetcenter.org/database.html) or [ESCO](https://esco.ec.europa.eu/), but this is optional.

---

## Objective

Create a prototype that can answer questions such as:

- "Which skills distinguish Machine Learning Engineer roles from Data Scientist roles?"
- "What skills most often co-occur with LLM, generative AI, or RAG-related requirements?"
- "Which companies hire for both data engineering and machine-learning-related roles, and what skills connect those roles?"
- "Given Python, SQL, Docker, and NLP experience, what adjacent roles are relevant and which skills are most commonly missing?"
- "What evidence supports the claim that production ML roles require cloud and software-engineering skills in addition to modelling experience?"

Answers should be grounded in the source data and include traceable evidence.

---

## Requirements

| Component | What | Deliverable |
|-----------|------|-------------|
| 1. Baseline RAG | Embeddings, vector store, retrieval, LLM answer with citations | **Working implementation** |
| 2. Knowledge Graph | Model entities and relationships from the structured data | **Working implementation** |
| 3. Graph-RAG | How graph structure would improve retrieval beyond vector search | **Architectural design** (written description or diagram) |

---

## 1. Baseline RAG

Implement a conventional RAG pipeline over job-posting text.

At minimum:

1. Create embeddings for job descriptions or appropriately sized chunks.
2. Index them in a vector store.
3. Retrieve relevant postings for a natural-language question.
4. Use an LLM to produce an answer based only on retrieved evidence.
5. Return citations or references to the underlying job postings.

You may use **FAISS**, Chroma, pgvector, or another appropriate local vector-search option.

For each cited posting, include enough information for a reviewer to identify it, for example:

- job ID;
- job title;
- company;
- location, where available;
- a short supporting excerpt.

A UI is not required. A notebook, CLI, Streamlit app, or simple API is sufficient.

---

## 2. Knowledge graph

Build a small knowledge graph from the structured data and/or information extracted from job descriptions.

You may use **NetworkX**, Kùzu, DuckDB/SQLite, Neo4j, RDFLib, or another justified approach. A graph database is not required.

### Suggested graph model

Possible node types:

- `JobPosting`
- `Company`
- `Skill`
- `Role`
- `Industry`
- `Location`
- `Occupation` — optional, e.g. from ESCO/O*NET

Possible edge types:

```text
(JobPosting) ──POSTED_BY──────> (Company)
(JobPosting) ──HAS_ROLE────────> (Role)
(JobPosting) ──MENTIONS_SKILL──> (Skill)
(JobPosting) ──IN_INDUSTRY─────> (Industry)
(JobPosting) ──LOCATED_IN──────> (Location)
(Company) ──OPERATES_IN────────> (Industry)
(Skill) ──RELATED_TO───────────> (Skill)       [optional/derived]
(Skill) ──RELEVANT_TO──────────> (Occupation)  [optional enrichment]
```

You do not need to implement every entity or relationship above. Choose a small, meaningful schema and document why it supports your questions.

---

## 3. Graph-RAG design

Describe how you would integrate graph structure into the retrieval pipeline to improve answers beyond what vector search alone can provide.

You do **not** need to implement this. Provide an architectural description (text, diagram, or pseudocode) that covers:

- Which question types benefit from graph-aware retrieval and why;
- How graph traversal or expansion would interact with vector retrieval;
- What the input/output of the graph step would be;
- What signals from the graph would improve answer quality.

Describe at least one concrete approach and explain why it would improve answers for a specific question type from this dataset.

If you have time to implement any part of Graph-RAG, that is welcome but not expected.

---

### Demonstration

Demonstrate your system on at least **three questions** covering:

1. **Direct retrieval** — e.g. "Which postings mention retrieval-augmented generation, vector databases, or LLM application development?"
2. **Comparison / synthesis** — e.g. "How do the requested skills differ between Machine Learning Engineer and Data Scientist roles in this corpus?"
3. **Knowledge graph query** — e.g. "Which companies hire for both data engineering and ML engineering roles, and which skills occur across both role types?" Show how your graph answers this (traversal, aggregation, or structured query).

For questions 1 and 2, show the working RAG output with citations. For question 3, show how your knowledge graph answers it, and describe how Graph-RAG would improve the answer if integrated into the retrieval pipeline.

---

# Deliverables

## 1. Code repository

Submit a GitHub repository or zipped project containing:

- source code;
- dependency specification (`requirements.txt`, `pyproject.toml`, etc.);
- reproducible setup and run instructions;
- a clear description of environment variables or API keys, if applicable.

The solution should be runnable locally. An LLM API (using your own key) is acceptable — document which provider/model is used and ensure the system degrades gracefully or provides a mock mode if the key is unavailable.

## 2. README / design note

Include a concise design note covering your architecture, key design decisions, and trade-offs. We care about clarity of thinking, not length.

## 3. Presentation

Prepare a **15-minute presentation** for the Data Science / ML team, followed by discussion.

A live demo is welcome but not required; prepared outputs are acceptable.

---
## Data files

The following data files are provided in the `data/` directory:

| File | Rows | Description |
|------|-----:|-------------|
| `sample_postings.csv` | 4,072 | Job postings with title, description, company, location, salary, experience level, etc. |
| `sample_companies.csv` | 1,979 | Company profiles with name, description, size, location, and URL |
| `sample_skills.csv` | 6,800 | Skill tags linked to job postings (job_id → skill_abr) |
| `sample_skill_lookup.csv` | 35 | Skill abbreviation → full skill name mapping |
| `sample_industries.csv` | 422 | Industry labels (industry_id, industry_name) |

### Key columns

**sample_postings.csv:** `job_id`, `company_name`, `title`, `description`, `location`, `company_id`, `formatted_work_type`, `formatted_experience_level`, `skills_desc`, `max_salary`, `min_salary`, `remote_allowed`

**sample_companies.csv:** `company_id`, `name`, `description`, `company_size`, `state`, `country`, `city`, `url`

**sample_skills.csv:** `job_id`, `skill_abr`

**sample_skill_lookup.csv:** `skill_abr`, `skill_name` — maps abbreviations (e.g. `ENGR`) to full names (e.g. `Engineering`)

**sample_industries.csv:** `industry_id`, `industry_name`

In [1]:
from jobmarket.config import DATA_DIR
from jobmarket.data import load_postings

postings = load_postings()
print(f"Postings: {len(postings):,}")
print(f"Columns: {list(postings.columns)}")


Postings: 4,072
Columns: ['job_id', 'company_name', 'title', 'description', 'max_salary', 'pay_period', 'location', 'company_id', 'views', 'med_salary', 'min_salary', 'formatted_work_type', 'applies', 'original_listed_time', 'remote_allowed', 'job_posting_url', 'application_url', 'application_type', 'expiry', 'closed_time', 'formatted_experience_level', 'skills_desc', 'listed_time', 'posting_domain', 'sponsored', 'work_type', 'currency', 'compensation_type', 'normalized_salary', 'zip_code', 'fips']


In [2]:
postings.description.values[10]

'Responsibilities:Develop new features, fix bugs, Writing and maintaining code in React, JavaScript, Typescript, and Node.js Using data and analytics to improve the product and customer experienceAccountable for code quality, including writing unit and integration testsWorking collaboratively with a team and refactor PHP and Javascript code for both the frontend and backend of a B2B SaaS product. You will work with another developer via Ping-Pong Programming to help you get up to speed on the code and be part of our initiative for Test Driven Development.Collaborate with the IT team to implement, test, and continuously deliver new features.Create and maintain endpoints.Contribute to all phases of the software development lifecycle.Design and develop applications in React, ensuring scalability, repeatability, support, etc.Develop complex, intelligent, and secure software applications with modern technologies for implementation on both client and servicer side.Build frontend UIs that are

In [3]:
# Load related tables
from jobmarket.data import load_companies, load_industries, load_skills

skills = load_skills()
companies = load_companies()
industries = load_industries()

if skills is not None:
    print(f"Skills entries: {len(skills):,}")
if companies is not None:
    print(f"Companies: {len(companies):,}")
if industries is not None:
    print(f"Industries: {len(industries):,}")


Skills entries: 6,800
Companies: 1,979
Industries: 422


## Explore the data

In [4]:
postings.head(10)

,job_id,company_name,title,description,max_salary,pay_period,location,company_id,views,med_salary,...,skills_desc,listed_time,posting_domain,sponsored,work_type,currency,compensation_type,normalized_salary,zip_code,fips
0,133130219,NaN,Software Engineer,"Education Bachelor's degree in software, math,...",NaN,NaN,Los Angeles Metropolitan Area,NaN,1.0,NaN,...,NaN,1.713535e+12,NaN,0,FULL_TIME,NaN,NaN,NaN,NaN,NaN
1,175485704,GOYT,Software Engineer,Job Description:GOYT is seeking a skilled and ...,NaN,NaN,"Denver, CO",76987056.0,273.0,NaN,...,NaN,1.713281e+12,NaN,0,PART_TIME,NaN,NaN,NaN,80202.0,8031.0
2,2234533717,Ideando Inc,Full Stack Engineer,"Location: Remote\nCompany Overview:SkillFit, a...",NaN,NaN,United States,69611476.0,21.0,NaN,...,NaN,1.713493e+12,NaN,0,FULL_TIME,NaN,NaN,NaN,NaN,NaN
3,2735151240,LT Apparel Group,Intern- Business Analytics,Summer Internshipadidas & Carhartt kids Divisi...,NaN,NaN,Greensboro--Winston-Salem--High Point Area,103060.0,5.0,NaN,...,NaN,1.712862e+12,NaN,0,TEMPORARY,NaN,NaN,NaN,NaN,NaN
4,3540371917,KeyBank,Enterprise Data & Analytics Infrastructure Man...,The Enterprise Data Infrastructure and Analyti...,NaN,NaN,"Cleveland, OH",3252.0,30.0,NaN,...,NaN,1.713452e+12,NaN,0,FULL_TIME,NaN,NaN,NaN,44102.0,39035.0
5,3586167732,StyleAI,Senior Software Engineer,"StyleAI is the AI-powered, all-in-one unified ...",NaN,NaN,San Francisco Bay Area,90662302.0,31.0,NaN,...,NaN,1.713397e+12,NaN,0,FULL_TIME,NaN,NaN,NaN,NaN,NaN
6,3728459637,HireBus,Principal Backend Engineer,Principal Backend Engineer - Join HireBus and ...,225.0,YEARLY,United States,91326370.0,7.0,NaN,...,NaN,1.713471e+12,NaN,0,FULL_TIME,USD,BASE_SALARY,212.5,NaN,NaN
7,3742692445,ZenithMinds Inc,Sr Data Engineer with Kafka,Data Engineer with Kafka (W2 Only)💯% Remote\nM...,NaN,NaN,"Austin, TX",81941852.0,39.0,NaN,...,NaN,1.713209e+12,NaN,0,FULL_TIME,NaN,NaN,NaN,78701.0,48453.0
8,3797449314,"Subaru Research and Development, Inc",Cloud Platform/ Big Data Engineer,About Subaru Research and Development:Do you c...,NaN,NaN,"Michigan, United States",91080439.0,7.0,NaN,...,NaN,1.713553e+12,NaN,0,FULL_TIME,NaN,NaN,NaN,NaN,NaN
9,3800272386,Sovrinti,Data Science Software Engineer,Company Overview Sovrinti is a full-service en...,NaN,NaN,United States,80936343.0,125.0,NaN,...,NaN,1.712347e+12,NaN,0,FULL_TIME,NaN,NaN,NaN,NaN,NaN


In [6]:
missing_company_id = postings[postings["company_id"].isna()]
print(missing_company_id)

          job_id company_name  \
0      133130219          NaN   
16    3835965860          NaN   
17    3841003722          NaN   
126   3884915616          NaN   
147   3885105100          NaN   
153   3885108780          NaN   
158   3885110295          NaN   
162   3885114238          NaN   
255   3886207306          NaN   
278   3886482903          NaN   
355   3887102746          NaN   
486   3888029592          NaN   
512   3888813149          NaN   
654   3890861573          NaN   
964   3894662724          NaN   
1025  3895200320          NaN   
1145  3898157558          NaN   
1167  3898160447          NaN   
1653  3901615813          NaN   
1709  3901689923          NaN   
1787  3901939171          NaN   
1797  3901940794          NaN   
1867  3901946817          NaN   
1941  3901979406          NaN   
1965  3901992635          NaN   
2012  3902343369          NaN   
2035  3902350998          NaN   
2378  3902906828          NaN   
2551  3903485166          NaN   
2765  3904

In [7]:
missing_company_id_with_name = postings[
    postings["company_id"].isna() & postings["company_name"].notna()
]
print(missing_company_id_with_name)

Empty DataFrame
Columns: [job_id, company_name, title, description, max_salary, pay_period, location, company_id, views, med_salary, min_salary, formatted_work_type, applies, original_listed_time, remote_allowed, job_posting_url, application_url, application_type, expiry, closed_time, formatted_experience_level, skills_desc, listed_time, posting_domain, sponsored, work_type, currency, compensation_type, normalized_salary, zip_code, fips]
Index: []

[0 rows x 31 columns]


In [8]:
company_id_conflicts = (
    postings
    .dropna(subset=["company_name"])               # ignore missing names
    .groupby("company_name")[
        "company_id"
    ]
    .nunique()
    .reset_index(name="unique_company_ids")
    .query("unique_company_ids > 1")
)

print(company_id_conflicts)

     company_name  unique_company_ids
440  Confidential                   4
603       Entegee                   2


In [9]:
duplicate_names = company_id_conflicts["company_name"].tolist()

conflicting_postings = postings[
    postings["company_name"].isin(duplicate_names)
].sort_values(["company_name", "company_id", "job_id"])

print(conflicting_postings[["job_id", "company_name", "company_id", "title", "location"]])

          job_id  company_name  company_id  \
3351  3905213470  Confidential   2183652.0   
776   3891252366  Confidential  14383281.0   
3129  3904922588  Confidential  18348487.0   
1061  3895508294  Confidential  33242739.0   
1063  3895518970  Confidential  33242739.0   
3497  3905306415  Confidential  33242739.0   
684   3890898200       Entegee     32090.0   
1252  3899523514       Entegee     32090.0   
1754  3901936750       Entegee     32090.0   
1459  3901345527       Entegee    530662.0   

                                                  title  \
3351              Senior Business Intelligence Director   
776   Senior C++ Software Engineer (Medical Devices)...   
3129         Senior Java Developer (Full Stack) - $120K   
1061                           Senior Software Engineer   
1063                        Principal Software Engineer   
3497                  Loan Transformation Data Analyst    
684                        Software Engineering Manager   
1252                 

In [10]:
# Role distribution
postings["title"].value_counts().head(20)

title
Software Engineer                   181
Senior Software Engineer            164
Data Analyst                        137
Data Engineer                        91
Full Stack Engineer                  58
DevOps Engineer                      57
Data Scientist                       56
Senior Data Engineer                 44
Cloud Engineer                       32
Embedded Software Engineer           30
Senior Data Analyst                  25
Machine Learning Engineer            23
Principal Software Engineer          21
Senior Data Scientist                21
Lead Software Engineer               20
Online Data Analyst                  20
Software Engineer in Test            20
Staff Software Engineer              19
Senior Machine Learning Engineer     19
Java Software Engineer               16
Name: count, dtype: int64

In [11]:
# Check description field
print(f"Postings with description: {postings['description'].notna().sum():,}")
print(f"\nSample description (first 500 chars):")
print(postings['description'].dropna().iloc[0][:500])

Postings with description: 4,072

Sample description (first 500 chars):
Education Bachelor's degree in software, math, or science required Job Skills Analytical skills, group work, knowledge of intended audience, understanding of different roles


In [12]:
if skills is not None:
    print("Top 20 skills:")
    skill_col = "skill_abr" if "skill_abr" in skills.columns else skills.columns[-1]
    print(skills[skill_col].value_counts().head(20))

Top 20 skills:
skill_abr
IT      3037
ENG     2126
ANLS     391
RSCH     259
OTHR     161
SALE      86
FIN       78
CNSL      78
BD        76
MGMT      68
MRKT      48
SCI       46
MNFC      36
STRA      35
PRDM      33
HR        29
DSGN      27
GENB      22
QA        22
ADM       21
Name: count, dtype: int64


In [13]:
if companies is not None:
    display(companies.head(10))

,company_id,name,description,company_size,state,country,city,zip_code,address,url
0,1009,IBM,"At IBM, we do more than work. We create. We cr...",7.0,NY,US,"Armonk, New York",10504,International Business Machines Corp.,https://www.linkedin.com/company/ibm
1,1016,GE HealthCare,Every day millions of people feel the impact o...,7.0,0,US,Chicago,0,-,https://www.linkedin.com/company/gehealthcare
2,1025,Hewlett Packard Enterprise,Official LinkedIn of Hewlett Packard Enterpris...,7.0,Texas,US,Houston,77389,1701 E Mossy Oaks Rd Spring,https://www.linkedin.com/company/hewlett-packa...
3,1028,Oracle,We’re a cloud technology company that provides...,7.0,Texas,US,Austin,78741,2300 Oracle Way,https://www.linkedin.com/company/oracle
4,1033,Accenture,Accenture is a leading global professional ser...,7.0,0,IE,Dublin 2,0,Grand Canal Harbour,https://www.linkedin.com/company/accenture
5,1035,Microsoft,Every company has a mission. What's ours? To e...,7.0,Washington,US,Redmond,98052,1 Microsoft Way,https://www.linkedin.com/company/microsoft
6,1038,Deloitte,Deloitte drives progress. Our firms around the...,7.0,0,OO,Worldwide,0,Worldwide,https://www.linkedin.com/company/deloitte
7,1044,PwC,"At PwC, our purpose is to build trust in socie...",7.0,0,GB,0,0,1 Embankment Place,https://www.linkedin.com/company/pwc
8,1052,AT&T,We understand that our customers want an easie...,7.0,TX,US,Dallas,75202,208 S. Akard Street,https://www.linkedin.com/company/att
9,1053,Intel Corporation,Intel’s mission is to shape the future of tech...,7.0,California,US,Santa Clara,95052,Robert Noyce Building,https://www.linkedin.com/company/intel-corpora...


In [14]:
if industries is not None:
    display(industries.head(10))

,industry_id,industry_name
0,1,Defense and Space Manufacturing
1,3,Computer Hardware Manufacturing
2,4,Software Development
3,5,Computer Networking Products
4,6,"Technology, Information and Internet"
5,7,Semiconductor Manufacturing
6,8,Telecommunications
7,9,Law Practice
8,10,Legal Services
9,11,Business Consulting and Services


---
## Your work starts here

Use the cells above as a starting point for exploration. Add new cells below to build your solution.

## 1. Baseline RAG — Implementation

Pipeline: chunk posting text → embed locally with `sentence-transformers` (no API key required) → index in two interchangeable vector stores (**FAISS** exact search and **turbovec** 4-bit quantized search) → retrieve top-k postings for a question → generate a grounded answer with Claude (`claude-haiku-4-5`) → return the answer with citations (job ID, title, company, location, excerpt).

If no Anthropic credentials are available (or a call fails), answer generation falls back to a deterministic, extractive mock answer built directly from the retrieved evidence, so the whole pipeline runs end-to-end without an API key.

In [ ]:
import anthropic

from jobmarket.config import (
    ANTHROPIC_MODEL,
    CHUNK_OVERLAP_WORDS,
    CHUNK_SIZE_WORDS,
    EMBEDDING_MODEL_NAME,
    TOP_K_CHUNKS,
    TOP_K_POSTINGS,
)

# Resolves ANTHROPIC_API_KEY if present; harmless if absent —
# absence/failure is handled at call time in generate_answer() below.
anthropic_client = anthropic.Anthropic()


### Chunk-size analysis

Before picking a chunk size, sanity-check it against the actual data rather than picking a round number blind.

**Why chunking is needed at all:** `description` (+ `skills_desc`) length varies hugely across postings — the distribution below shows a median around ~490 words, so most postings would be truncated (or leave most of their content unembedded) if each posting were indexed as a single vector.

**Why 180 words / 40-word overlap specifically:** the embedding model (`all-MiniLM-L6-v2`) has a hard **256-token** max sequence length — text beyond that is truncated silently by the tokenizer, not an error. 180 words is meant to sit comfortably under that limit, leaving headroom for the `Job title: ... / Company: ... / Location: ...` prefix that `build_corpus` prepends to every chunk before embedding (see `embedding_text` in [chunking.py](jobmarket/rag/chunking.py)). The 40-word overlap (~22% of the chunk) is a conventional fraction used to preserve context across chunk boundaries without materially inflating the number of chunks (and therefore index size / embedding cost).

In [47]:
import numpy as np
from transformers import AutoTokenizer

from jobmarket.rag.chunking import split_into_chunks

MAX_SEQ_TOKENS = 256  # all-MiniLM-L6-v2's max_seq_length

# --- word-length distribution: does chunking make sense for this corpus at all? ---
body_text = (
    postings["description"].fillna("") + " " + postings["skills_desc"].fillna("")
).str.strip()
word_counts = body_text.str.split().apply(len)

print("description(+skills_desc) length in words:")
print(word_counts.describe(percentiles=[.25, .5, .75, .9, .95, .99]).round(1))
print(f"\nPostings longer than a single {CHUNK_SIZE_WORDS}-word chunk: "
      f"{(word_counts > CHUNK_SIZE_WORDS).mean():.1%}")

# --- token-budget check: word count is a proxy for token count, not the real constraint ---
tokenizer = AutoTokenizer.from_pretrained(EMBEDDING_MODEL_NAME)

sample_chunks = [
    chunk
    for text in body_text.sample(200, random_state=42)
    for chunk in split_into_chunks(text, CHUNK_SIZE_WORDS, CHUNK_OVERLAP_WORDS)
]
chunk_token_counts = np.array([len(tokenizer(c)["input_ids"]) for c in sample_chunks])

# embedding_text also prepends "Job title: ...\nCompany: ...\nLocation: ...\n\n" before the chunk text
prefix_tokens = len(tokenizer("Job title: X\nCompany: Y\nLocation: Z\n\n")["input_ids"])

print(f"\nTokenizer: {EMBEDDING_MODEL_NAME}, model max_seq_length={MAX_SEQ_TOKENS}")
print(f"Chunk token counts (n={len(chunk_token_counts)} chunks from {CHUNK_SIZE_WORDS}-word splitting): "
      f"mean={chunk_token_counts.mean():.0f}, p90={np.percentile(chunk_token_counts, 90):.0f}, "
      f"max={chunk_token_counts.max()}")
print(f"Chunks exceeding {MAX_SEQ_TOKENS} tokens (silently truncated by the model): "
      f"{(chunk_token_counts > MAX_SEQ_TOKENS).mean():.1%}")
print(f"Citation-prefix overhead added on top in embedding_text: ~{prefix_tokens} tokens")


description(+skills_desc) length in words:
count    4072.0
mean      532.0
std       327.4
min         1.0
25%       253.0
50%       489.0
75%       755.0
90%       978.0
95%      1113.0
99%      1350.3
max      2486.0
dtype: float64

Postings longer than a single 180-word chunk: 85.4%

Tokenizer: sentence-transformers/all-MiniLM-L6-v2, model max_seq_length=256
Chunk token counts (n=814 chunks from 180-word splitting): mean=221, p90=270, max=390
Chunks exceeding 256 tokens (silently truncated by the model): 19.3%
Citation-prefix overhead added on top in embedding_text: ~12 tokens


**Measured on this dataset:** median `description`+`skills_desc` length is ~489 words (mean 532, up to 2,486) and 85% of postings exceed a single 180-word chunk, confirming multi-chunk splitting is necessary for most postings. Tokenizing 180-word chunks with the actual `all-MiniLM-L6-v2` tokenizer, however, shows word count is an imperfect proxy for token count: chunks average ~221 tokens (not 180), and roughly **1 in 5 chunks (19%) exceeds the 256-token limit** and has its tail silently truncated by the model — before even adding the ~12-token citation prefix.

This is accepted as a minor trade-off rather than fixed by shrinking the chunk size further: the 40-word overlap means a truncated tail is largely re-covered as the *start* of the next chunk, so no content is permanently lost from retrieval — only, occasionally, scored from a slightly shorter effective window. `CHUNK_SIZE_WORDS` / `CHUNK_OVERLAP_WORDS` stay at 180/40.

### Chunking

`description` (and `skills_desc`, where present) is unstructured free text of varying length. Each posting's text is split into overlapping ~180-word chunks (40-word overlap, rationale and token-budget analysis above) so long postings aren't truncated and retrieval can point at the relevant part of a posting rather than the whole document. Postings with no usable text are skipped. Each chunk keeps a back-reference to its parent posting (`job_id`, `title`, `company`, `location`) for citations.

In [48]:
from jobmarket.rag.chunking import build_corpus, split_into_chunks

corpus = build_corpus(postings, chunk_size=CHUNK_SIZE_WORDS, overlap=CHUNK_OVERLAP_WORDS)
print(f"Built {len(corpus):,} chunks from {corpus['job_id'].nunique():,} postings "
      f"(out of {len(postings):,} total postings)")


Built 16,378 chunks from 4,072 postings (out of 4,072 total postings)


Note: 
A framework such as LangChain, provudes implementation for several chunking strategies. See: https://docs.langchain.com/oss/python/integrations/splitters 

### Embeddings

Chunks are embedded locally with `all-MiniLM-L6-v2` (384-dim, `sentence-transformers`) — no external API key required, so index-building always works offline. Embeddings are L2-normalized so inner-product search is equivalent to cosine similarity.

In [49]:
from sentence_transformers import SentenceTransformer

embedder = SentenceTransformer(EMBEDDING_MODEL_NAME)

embeddings = embedder.encode(
    corpus["embedding_text"].tolist(),
    batch_size=64,
    show_progress_bar=True,
    normalize_embeddings=True,
    convert_to_numpy=True,
).astype("float32")

EMBEDDING_DIM = embeddings.shape[1]
print(f"Embedded {embeddings.shape[0]:,} chunks into {EMBEDDING_DIM}-dim vectors")

Loading weights:   0%|          | 0/103 [00:00<?, ?it/s]

Batches:   0%|          | 0/256 [00:00<?, ?it/s]

Embedded 16,378 chunks into 384-dim vectors


Note:
A framework like LangChain provide multiple embedding models integration, as well as, guidelines and metrics for selecting the best embedding model for the use case.
URL: https://docs.langchain.com/oss/python/integrations/embeddings

### Vector store: FAISS and turbovec

Two interchangeable local vector-search backends sit behind a common `VectorStore.search()` interface:

- **FAISS** (`IndexFlatIP`) — exact brute-force cosine search via normalized inner product. Simple and exact, the natural baseline at this corpus size (a few thousand chunks).
- **turbovec** (`IdMapIndex`, 4-bit quantized) — a SIMD, quantized index built for corpora far larger than this one (millions of vectors); included as a second approach, to compare retrieval quality/latency against the exact FAISS index at small scale, where the quantization trade-off doesn't pay for itself yet.

Note:
A framework like LangChain provides several vector store integration options.
URL: https://docs.langchain.com/oss/python/integrations/vectorstores

In [50]:
from jobmarket.rag.vector_store import build_vector_stores

vector_stores = build_vector_stores(embeddings)
print("Vector stores ready:", list(vector_stores.keys()))


Vector stores ready: ['faiss', 'turbovec']


### Retrieval

A query is embedded with the same model, searched against the chosen backend, then chunks are mapped back to their source posting and de-duplicated by `job_id` (keeping each posting's best-scoring chunk) so citations point at distinct postings rather than repeating one posting multiple times.

In [51]:
from jobmarket.rag.retrieval import retrieve as _retrieve


def retrieve(
    query: str,
    k_chunks: int = TOP_K_CHUNKS,
    k_postings: int = TOP_K_POSTINGS,
    backend: str = "faiss",
) -> list[dict]:
    return _retrieve(
        query,
        corpus=corpus,
        vector_stores=vector_stores,
        embedder=embedder,
        k_chunks=k_chunks,
        k_postings=k_postings,
        backend=backend,
    )


### Answer generation

Retrieved postings are formatted as a numbered evidence list and passed to Claude (`claude-haiku-4-5`, chosen for speed/cost during iterative development — see design note) with instructions to answer using only that evidence and to cite postings by their bracket number. If no Anthropic credentials are available, or the call fails for any reason (network, auth, rate limit), `generate_answer` falls back to a deterministic, extractive mock answer built directly from the retrieved excerpts — the "mock mode" called for by the assignment, so the pipeline always runs end-to-end without an API key.

In [52]:
from jobmarket.rag.generation import SYSTEM_PROMPT, format_evidence, mock_answer
from jobmarket.rag.generation import generate_answer as _generate_answer


def generate_answer(query: str, evidence: list[dict], model: str = ANTHROPIC_MODEL) -> str:
    return _generate_answer(query, evidence, anthropic_client=anthropic_client, model=model)


In [53]:
def answer_question(
    query: str,
    backend: str = "faiss",
    k_chunks: int = TOP_K_CHUNKS,
    k_postings: int = TOP_K_POSTINGS,
    model: str = ANTHROPIC_MODEL,
    verbose: bool = True,
) -> dict:
    evidence = retrieve(query, k_chunks=k_chunks, k_postings=k_postings, backend=backend)
    answer = generate_answer(query, evidence, model=model)

    if verbose:
        print(f"Q: {query}\n(backend={backend})\n")
        print(answer)
        print("\nCitations:")
        for i, e in enumerate(evidence, start=1):
            print(f"  [{i}] job_id={e['job_id']} | {e['title']} | {e['company']} | {e['location']} | score={e['score']:.3f}")
            print(f"      \"{e['excerpt'][:160].strip()}...\"")

    return {"query": query, "answer": answer, "citations": evidence, "backend": backend}

### Demonstration

Two of the three required demo question types are covered here — the third (knowledge-graph query) is addressed in Part 2. A backend comparison (FAISS vs. turbovec) is included at the end.

#### Direct retrieval

In [54]:
_ = answer_question(
    "Which postings mention retrieval-augmented generation, vector databases, or LLM application development?",
    backend="faiss",
)

Q: Which postings mention retrieval-augmented generation, vector databases, or LLM application development?
(backend=faiss)

# Postings Mentioning Retrieval-Augmented Generation, Vector Databases, or LLM Application Development

Based on the provided evidence, I can identify postings that mention LLM application development, but **none explicitly mention retrieval-augmented generation or vector databases**:

**Postings with LLM application development:**

- [1] **LLM Data Scientist — SS&C Technologies**: Mentions "Large Language Model (LLM) Expertise: Leverage your expertise in working with large language models" and advanced NLP model development.

- [2] **Sr. Software Engineer, Search Relevance — Moveworks**: While not explicitly calling it RAG, this posting describes building "search based question answering platform" that combines "information retrieval methodology with the latest advancement in natural language understanding" and involves "accurately retrieving, matching and ranki

#### Comparison / synthesis

In [55]:
_ = answer_question(
    "Which skills distinguish Machine Learning Engineer roles from Data Scientist roles?",
    backend="faiss",
)

Q: Which skills distinguish Machine Learning Engineer roles from Data Scientist roles?
(backend=faiss)

# Skills That Distinguish Machine Learning Engineers from Data Scientists

Based on the evidence provided, the key distinguishing skills are:

## Machine Learning Engineer Skills:

**Implementation and Production Focus:**
- [1] Deploy machine learning models into production environments and work closely with DevOps teams to ensure scalability, reliability, and security
- [5] Design and implement infrastructure and platform components for training, deploying, and monitoring machine learning models in production
- [5] Build pipelines to ingest data from myriad sources into a centralized data lake

**Model Operations and Monitoring:**
- [1] Implement monitoring solutions to track model performance in real-time and continuously fine-tune and optimize models

**Leadership in Implementation:**
- [4] The manager is "looking for a Machine Learning engineer that can truly implement/lead. Trad

In [56]:
_ = answer_question(
    "Given Python, SQL, Docker, and NLP experience, what adjacent roles are relevant "
    "and which skills are most commonly missing?",
    backend="faiss",
)

Q: Given Python, SQL, Docker, and NLP experience, what adjacent roles are relevant and which skills are most commonly missing?
(backend=faiss)

# Adjacent Roles and Missing Skills Analysis

Based on the evidence provided, here are the relevant adjacent roles and commonly missing skills:

## Adjacent Relevant Roles

1. **Senior Lead Software Engineer** (ML/Data focus) [4] — Combines your Python, SQL, and Docker expertise with machine learning specialization
2. **Lead Python Backend Engineer** [1][2] — Leverages your Python and SQL skills with cloud infrastructure focus
3. **Senior Software Engineer (ETL/Data)** [3][5] — Utilizes SQL expertise in data pipeline roles

## Most Commonly Missing Skills

Based on job requirements across these roles:

1. **Cloud Platforms** — AWS environment experience is heavily emphasized [1][2], appearing as a core requirement
2. **Machine Learning Frameworks** — TensorFlow, PyTorch, or Scikit-learn [4] represent a major gap for NLP-adjacent roles
3. **Data

#### FAISS vs. turbovec — backend comparison

Both vector-search backends sit behind the same interface, so any question can be run against either. Below, the same query is retrieved from both to sanity-check that the quantized turbovec index returns materially the same top postings as the exact FAISS index.

In [57]:
comparison_query = "What skills most often co-occur with LLM, generative AI, or RAG-related requirements?"

for backend in ("faiss", "turbovec"):
    print(f"=== backend: {backend} ===")
    result = answer_question(comparison_query, backend=backend, verbose=False)
    for i, e in enumerate(result["citations"], start=1):
        print(f"  [{i}] job_id={e['job_id']} | {e['title']} | {e['company']} | score={e['score']:.3f}")
    print()

=== backend: faiss ===
  [1] job_id=3884432812 | Generative AI Engineer | Ehub Global Inc | score=0.513
  [2] job_id=3888994426 | 【For External Referral - Talent Spotters】Research Scientist for Generative AI, Multimodal and LLM | TikTok | score=0.509
  [3] job_id=3904938734 | LLM Data Scientist | SS&C Technologies | score=0.501
  [4] job_id=3899532438 | Data Scientist - GenAI - Solution Specialist - Hybrid | Deloitte | score=0.495
  [5] job_id=3902868262 | Principal Applied AI Engineer | Alaffia Health | score=0.494

=== backend: turbovec ===
  [1] job_id=3884432812 | Generative AI Engineer | Ehub Global Inc | score=0.510
  [2] job_id=3888994426 | 【For External Referral - Talent Spotters】Research Scientist for Generative AI, Multimodal and LLM | TikTok | score=0.508
  [3] job_id=3904938734 | LLM Data Scientist | SS&C Technologies | score=0.501
  [4] job_id=3899532438 | Data Scientist - GenAI - Solution Specialist - Hybrid | Deloitte | score=0.499
  [5] job_id=3902868262 | Principal App

## 2. Knowledge Graph — Implementation

Built with **NetworkX** (`DiGraph`) — no server, easy to inspect in the notebook, and more than sufficient for a few thousand postings.

### Schema

| Node type | Key | Source |
|---|---|---|
| `JobPosting` | `job_id` | `sample_postings.csv` |
| `Company` | normalized `company_name` | `sample_postings.csv`, enriched from `sample_companies.csv` when `company_id` joins |
| `Role` | canonical role name | derived from `title` via a small keyword taxonomy |
| `Skill` | canonical skill name | extracted from `title` + `description` + `skills_desc` via keyword matching |
| `Location` | raw `location` string | `sample_postings.csv` |

Edges: `(JobPosting)-POSTED_BY->(Company)`, `(JobPosting)-HAS_ROLE->(Role)`, `(JobPosting)-MENTIONS_SKILL->(Skill)`, `(JobPosting)-LOCATED_IN->(Location)`.

**Deliberately not implemented:**
- `Industry` — neither `sample_postings.csv` nor `sample_companies.csv` carries an `industry_id` to join against `sample_industries.csv` in this sample, so there's no reliable edge to draw.
- `Occupation` (ESCO/O*NET) — optional enrichment, out of scope.
- `Skill-RELATED_TO-Skill` — not needed by the target questions; the required demo question is answered directly via posting-level traversal and set aggregation, no derived skill-similarity edges required.

**Key design trade-offs:**
- **Skills are *not* `sample_skills.csv`'s categories** (`IT`, `ENG`, `ANLS`, …) — those are too coarse to answer skill-level questions (noted during initial data exploration above). Instead, `Skill` nodes come from matching a curated vocabulary of ~45 concrete tech/data skills against posting text via case-insensitive keyword matching with symbol-aware boundaries (handles `C++`, `CI/CD`, `Node.js`, etc.). This is plain regex, not NER — fast and transparent, with an accepted false-negative/positive rate on ambiguous short tokens (documented at the vocabulary definition below).
- **`Role` is derived from `title`** using an ordered keyword taxonomy (most-specific pattern wins), covering the ~10 canonical roles that dominate this corpus (Data Engineer, Machine Learning Engineer, Data Scientist, DevOps Engineer, Software Engineer, …) — this directly supports the ML-Engineer-vs-Data-Scientist and data-engineering/ML-engineering-overlap questions.
- **`Company` is keyed by normalized name, not `company_id`**, since `company_id` is frequently `NaN` in the sample (per the earlier exploration cells) — this trades perfect de-duplication (name variants could fragment one real company into multiple nodes) for actually connecting most postings to a company node.
- **`Location` is the raw string, unnormalized** (e.g. "Boston, MA" and a hypothetical "Boston, Massachusetts" would be distinct nodes) — a scoping decision, not implemented here.

In [58]:
from collections import Counter

import networkx as nx

from jobmarket.kg.taxonomy import ROLE_PATTERNS, ROLE_TAXONOMY, classify_role, compile_alternation


In [59]:
# Curated tech/data skill vocabulary. Deliberately excludes bare single-word tokens with high
# false-positive risk in prose (e.g. "R", "Go", "node") — only unambiguous aliases are listed.
from jobmarket.kg.taxonomy import SKILL_PATTERNS, SKILL_VOCABULARY, extract_skills


### Build the graph

In [60]:
from jobmarket.kg.build import build_graph, normalize_company_key

kg = build_graph(postings, companies)
print(f"Graph built: {kg.number_of_nodes():,} nodes, {kg.number_of_edges():,} edges")


Graph built: 6,757 nodes, 33,351 edges


In [61]:
kinds = Counter(nx.get_node_attributes(kg, "kind").values())
edge_types = Counter(d["type"] for _, _, d in kg.edges(data=True))

print("Nodes by type:")
for k, v in kinds.most_common():
    print(f"  {k}: {v:,}")

print("\nEdges by type:")
for k, v in edge_types.most_common():
    print(f"  {k}: {v:,}")

role_counts = Counter(kg.nodes[v]["name"] for _, v, d in kg.edges(data=True) if d["type"] == "HAS_ROLE")
total_postings = sum(role_counts.values())
print("\nRole coverage:")
for role, count in role_counts.most_common():
    print(f"  {role}: {count:,} ({count / total_postings:.1%})")

skill_freq: Counter = Counter()
for _, v, d in kg.edges(data=True):
    if d["type"] == "MENTIONS_SKILL":
        skill_freq[kg.nodes[v]["name"]] += 1

print("\nTop 15 extracted skills by posting mentions:")
for skill, count in skill_freq.most_common(15):
    print(f"  {skill}: {count:,}")


Nodes by type:
  JobPosting: 4,072
  Company: 1,974
  Location: 646
  Skill: 54
  Role: 11

Edges by type:
  MENTIONS_SKILL: 21,172
  HAS_ROLE: 4,072
  LOCATED_IN: 4,072
  POSTED_BY: 4,035

Role coverage:
  Software Engineer: 1,193 (29.3%)
  Other: 998 (24.5%)
  Data Analyst: 410 (10.1%)
  Data Engineer: 364 (8.9%)
  Full Stack Engineer: 354 (8.7%)
  Data Scientist: 244 (6.0%)
  DevOps Engineer: 191 (4.7%)
  Machine Learning Engineer: 128 (3.1%)
  Cloud Engineer: 109 (2.7%)
  Embedded Software Engineer: 56 (1.4%)
  QA / Test Engineer: 25 (0.6%)

Top 15 extracted skills by posting mentions:
  Python: 1,837
  SQL: 1,642
  AWS: 1,192
  Agile: 1,180
  Java: 1,018
  Azure: 908
  Machine Learning: 882
  CI/CD: 724
  JavaScript: 655
  Kubernetes: 572
  GCP: 560
  Docker: 530
  Statistics: 510
  C++: 490
  Git: 463


### Graph queries

Small traversal helpers over `kg`, composed to answer the required knowledge-graph question by **traversal** (posting → company, posting → skill) plus **aggregation** (set intersection across two role-filtered subgraphs) — no query language needed at this scale.

In [62]:
from jobmarket.kg.queries import company_for_posting as _company_for_posting
from jobmarket.kg.queries import postings_by_company_for_role as _postings_by_company_for_role
from jobmarket.kg.queries import postings_for_role as _postings_for_role
from jobmarket.kg.queries import skills_for_posting as _skills_for_posting
from jobmarket.kg.queries import skills_for_role as _skills_for_role


def postings_for_role(role: str) -> list[str]:
    return _postings_for_role(kg, role)


def company_for_posting(posting_node: str) -> str | None:
    return _company_for_posting(kg, posting_node)


def skills_for_posting(posting_node: str) -> set[str]:
    return _skills_for_posting(kg, posting_node)


def postings_by_company_for_role(role: str) -> dict[str, list[str]]:
    return _postings_by_company_for_role(kg, role)


def skills_for_role(role: str) -> Counter:
    return _skills_for_role(kg, role)


In [63]:
from jobmarket.kg.queries import companies_bridging_roles as _companies_bridging_roles


def companies_bridging_roles(role_a: str, role_b: str, top_skills: int = 10) -> dict:
    return _companies_bridging_roles(kg, role_a, role_b, top_skills=top_skills)


### Demonstration — Knowledge-graph query

"Which companies hire for both data engineering and ML engineering roles, and which skills occur across both role types?"

In [64]:
result = companies_bridging_roles("Data Engineer", "Machine Learning Engineer")

print(
    f"'{result['role_a']}' postings appear at {result['num_companies_role_a']} companies; "
    f"'{result['role_b']}' postings appear at {result['num_companies_role_b']} companies.\n"
)
print(f"Companies hiring for BOTH roles ({len(result['bridging_companies'])}):\n")
for c in result["bridging_companies"]:
    print(f"- {c['company']}")
    print(f"    {result['role_a']} job_ids: {c['postings_a']}")
    print(f"    {result['role_b']} job_ids: {c['postings_b']}")
    shared = ", ".join(c["shared_skills"]) if c["shared_skills"] else "(no shared skills extracted)"
    print(f"    Shared skills at this company: {shared}")
    print()

print(f"Skills occurring across BOTH role types overall (ranked by combined mention count):")
for skill in result["shared_skills_overall"]:
    print(f"  - {skill}")

'Data Engineer' postings appear at 253 companies; 'Machine Learning Engineer' postings appear at 99 companies.

Companies hiring for BOTH roles (12):

- Akkodis
    Data Engineer job_ids: [3894213406, 3894299338, 3894400441]
    Machine Learning Engineer job_ids: [3901166848, 3901803271]
    Shared skills at this company: Java, Machine Learning, Python, SQL, Spark, Statistics

- Booz Allen Hamilton
    Data Engineer job_ids: [3903846324, 3903846326]
    Machine Learning Engineer job_ids: [3894616657]
    Shared skills at this company: Agile, Machine Learning

- Capital One
    Data Engineer job_ids: [3885805686, 3885806687, 3885808491, 3885810350, 3885810353, 3894291104, 3902782722, 3902785272, 3902786201, 3903802633, 3904049996, 3904052744, 3904054114, 3904574894, 3904574897, 3904574898, 3904575804, 3904578558, 3904581374, 9900000, 9900001, 9900002]
    Machine Learning Engineer job_ids: [3902780915, 3902780922, 3902780923, 3902780924, 3902781837]
    Shared skills at this company: AW

*How Graph-RAG would improve this answer if integrated into the retrieval pipeline: see Part 3 below.*

## 3. Graph-RAG Design

*Architectural description only — not implemented. The knowledge graph built in Part 2 already provides everything a real implementation would traverse; this section describes how it would plug into the Part 1 RAG pipeline.*

### Where baseline RAG alone falls short

Evidence from the demos already run above:

- **"Given Python, SQL, Docker, and NLP experience, what adjacent roles are relevant?"** — the RAG-only answer surfaced a *"Senior Software Engineer-(Mainframe/SQL/ETL)"* posting alongside genuinely adjacent roles. Vector similarity matched on shared keywords (SQL, generic tech-stack overlap) with no notion of "how close is this role, structurally, to the input skill profile" — there's no set-based or frequency-based reasoning in embedding space.
- **"Which skills distinguish ML Engineer from Data Scientist roles?"** — the RAG-only answer returned five individually-relevant chunks, but a genuine *difference* between two role categories is a comparison of two full frequency distributions, not something a handful of retrieved excerpts can establish reliably.
- More generally: vector top-k is a **sample** (k≈5–8 chunks out of 16,378 in this corpus). Any question requiring exhaustive coverage — "which companies do X **and** Y," "what fraction of postings mention Z" — cannot be answered correctly by sampling, no matter how good the embedding model is.

### Which question types benefit from graph-aware retrieval, and why

| Question type | Example (from this assignment) | Best mechanism | Why |
|---|---|---|---|
| Multi-hop entity intersection | *"Which companies hire for both data engineering and ML engineering roles?"* | Graph traversal + set intersection (**implemented in Part 2**) | Needs two role-filtered subsets intersected by `Company` — vector similarity has no native intersection operator |
| Exhaustive aggregation / co-occurrence | *"What skills most often co-occur with LLM, generative AI, or RAG requirements?"* | Graph: shared-neighbor counting over `MENTIONS_SKILL` edges | Needs corpus-wide counts; `Skill` nodes already exist for `Large Language Models`, `Generative AI`, `Retrieval-Augmented Generation` — co-occurring skills are just the other `Skill` neighbors of the same `JobPosting` nodes |
| Structured skill-profile comparison | *"Which skills distinguish ML Engineer from Data Scientist roles?"* | Graph: per-`Role` skill-frequency profile, then relative-frequency diff | "Distinguishing" is a comparison of two aggregated distributions, not a similarity search |
| Nearest-role / skill-gap reasoning | *"Given Python, SQL, Docker, NLP, what adjacent roles are relevant and what's missing?"* | Graph: rank `Role`s by skill-profile overlap with the input set, diff the top role's profile against the input | A role's skill profile is a structured aggregate the graph already computes (`skills_for_role`); posting-level text search can't isolate that role-level signal from noise |
| Open-ended semantic / narrative discovery | *"Which postings mention retrieval-augmented generation, vector databases, or LLM application development?"* | Vector RAG (**as implemented in Part 1**) | Free phrasing, no fixed taxonomy required — exactly what embeddings are for |
| Grounded quoting for citations | any answer needing a supporting excerpt | Vector RAG / posting text (**as implemented in Part 1**) | The graph tells you *which* postings qualify; only the source text supplies a quotable excerpt |

Most real questions are a mix of the top and bottom halves of this table — which is why the proposed design **combines** graph traversal and vector retrieval rather than replacing one with the other.

### Proposed architecture

```text
                       ┌───────────────────────┐
  natural-language ───▶│ 1. Query parsing       │
  question             │   role/skill/company/  │
                       │   location + intent op │
                       └───────────┬────────────┘
                                   │ structured intent
                                   ▼
                       ┌───────────────────────┐
                       │ 2. Graph traversal     │
                       │   (kg: exact,          │
                       │    exhaustive)         │
                       └───────────┬────────────┘
                candidate job_ids + aggregate stats
                       + traversal provenance
                                   │
                                   ▼
                       ┌───────────────────────┐
                       │ 3. Vector retrieval    │
                       │   restricted/boosted   │
                       │   to the candidate set │
                       └───────────┬────────────┘
                        quotable excerpts, scored
                                   │
                                   ▼
                       ┌───────────────────────┐
                       │ 4. Answer generation   │
                       │   (graph facts +       │
                       │    vector excerpts)    │
                       └───────────────────────┘
```

**Stage 1 — Query parsing.** Reuse `classify_role` / `extract_skills` (already built in Part 2) plus lightweight entity matching for company/location names to turn the question into a structured intent: `{roles: [...], skills: [...], companies: [...], locations: [...], op: "intersect" | "co_occur" | "diff" | "rank" | "open"}`. An LLM call (same model as answer generation) is the pragmatic way to classify `op` and resolve entities that don't match the fixed vocabularies exactly (e.g. "ML engineering roles" → `Role: Machine Learning Engineer`).

**Stage 2 — Graph traversal.** Run the structured intent against `kg` using the same traversal primitives from Part 2 (`postings_for_role`, `postings_by_company_for_role`, `skills_for_role`, `companies_bridging_roles`) or natural extensions of them (e.g. a `skill_cooccurrence(skill)` that walks `MENTIONS_SKILL` backward to `JobPosting` then forward to other `Skill`s). Output:
- a **candidate `job_id` set** (exact, not sampled) to hand to vector search;
- **aggregate statistics** computed over the whole graph (counts, intersections, differences) — not approximated from k retrieved chunks;
- **traversal provenance** — which edges justified each result (e.g. "Capital One qualifies via 22 `Data Engineer` postings and 5 `Machine Learning Engineer` postings") — citation-quality explainability a similarity score alone can't provide.

**Stage 3 — Graph-guided vector retrieval.** Instead of searching the full ~16K-chunk index, restrict `retrieve()` to the graph-derived candidate `job_id`s — an *allowlist* filter, the same hybrid pattern `turbovec`'s `IdMapIndex.search()` already supports natively (candidates narrowed by an external system, then ranked by embedding similarity within that set). This gets the best of both: exact structural correctness from the graph, plus a natural-language excerpt worth quoting from the vector index. Where the graph traversal alone already answers the question numerically (counts, company lists), this stage's only job is finding *one good supporting quote per fact*, not re-discovering which postings are relevant.

**Stage 4 — Answer generation.** Same `generate_answer()` from Part 1, but the evidence block passed to the LLM now has two tiers: **graph facts** (exact counts, set memberships, comparisons — computed, not estimated) and **vector excerpts** (quotable text, tied to specific `job_id`s for citation). This lets the model make statistically-grounded claims ("14 companies post both role types; Python and SQL are the two most common shared skills, appearing in 9 of them") backed by a real excerpt, instead of inferring a pattern from 5 retrieved snippets.

### What the graph step contributes

**Input:** the structured intent from Stage 1 (roles/skills/companies/locations + operation type).

**Output:**
1. A candidate `job_id` set — exact and exhaustive, not a top-k sample — that narrows and improves the precision of the vector-search stage.
2. Aggregate statistics (frequency counts, set intersections/differences, co-occurrence) that vector similarity cannot compute at all.
3. Traversal provenance for each result, giving the answer generator (and the eventual reader) an auditable "why" beyond a similarity score.

**Signals that improve answer quality specifically:**
- **Exhaustiveness** — corpus-wide counts instead of a k-sized sample, directly fixing the "distinguishing skills" and "co-occurrence" question types.
- **Structural operations** (intersection, difference, ranking) that have no equivalent in embedding space.
- **Canonicalization** — `Role` and `Skill` nodes collapse title/skill variants (e.g. "ML Engineer", "Machine Learning Engineer", "Senior Machine Learning Engineer" → one `Role` node) that vector search would otherwise treat as separate, loosely-related points in embedding space.
- **Provenance** — a traversal path is a stronger citation than a cosine-similarity score.

### Concrete walkthrough

**"Which companies hire for both data engineering and ML engineering roles, and what skills connect those roles?"** — the required demo question, already answered purely by graph traversal in Part 2. A full Graph-RAG pipeline would improve on that pure-graph answer in one specific way: **Stage 3** would attach a real supporting quote to each bridging company — e.g. for Capital One, retrieve the single best-matching excerpt from its Data Engineer and ML Engineer postings that best evidences "Python" or "Spark" — turning the current company/skill list into a fully cited answer, the same citation quality as the Part 1 RAG demos.

**"Given Python, SQL, Docker, and NLP experience, what adjacent roles are relevant, and which skills are most commonly missing?"** shows the gap more starkly. The Part 1 RAG-only demo surfaced a *Mainframe/SQL/ETL* posting as a top match — a false positive driven by superficial keyword overlap. A graph-aware version would instead: (1) build the input skill set `{Python, SQL, Docker, NLP}`; (2) for every `Role`, compute `skills_for_role(role)` and rank roles by overlap (e.g. Jaccard similarity) with the input set — `Machine Learning Engineer` and `Data Engineer` would score highest, generic `Software Engineer` moderately, and the Mainframe posting's actual role would rank far lower once measured against its *full* skill profile rather than one chunk's text; (3) diff the top-ranked role's aggregate skill profile against the input set to report the *missing* skills (e.g. "Machine Learning Engineer postings in this corpus also commonly require: AWS, Kubernetes, Spark — Docker is already covered"); (4) hand the top 1–2 qualifying postings per recommended role to vector retrieval for a quotable excerpt. This directly fixes the specific failure mode observed in the Part 1 demo.